File: src/Preferences.cs
using System.Text.Json;
using System.Runtime.InteropServices;

namespace jammer
{
    public class Preferences
    {
        string FileContent = @"
;Do not use characters outside ascii, if you use it needs to use the same encoding as csharp uses by default. ö = oem7, ä = oem3 etc...
;See https://learn.microsoft.com/en-us/dotnet/api/system.consolekey?view=net-8.0 for allowed characters.
;When using numbers, 'd' part is not needed
; Allowed modifiers are ctrl, shift, alt, ctrl + shift and ctrl + alt
;
[Keybinds]
PlayPause = Spacebar
Quit = Q
NextSong = N
PreviousSong = P
PlaySong = Shift + p
Forward5s = RightArrow
Backwards5s = LeftArrow
VolumeUp = UpArrow
VolumeDown = DownArrow
Shuffle = S
SaveAsPlaylist = Shift + Alt + S
SaveCurrentPlaylist = Shift + S
ShufflePlaylist = Alt + S
Loop = L
Mute = M
ShowHidePlaylist = F
ListAllPlaylists = Shift + F
Help = H
Settings = C
ToSongStart = 0
ToSongEnd = 9
PlaylistOptions = F2
ForwardSecondAmount = 1
BackwardSecondAmount = 2
ChangeVolumeAmount = 3
Autosave = 4
CurrentState = F12
CommandHelpScreen = Tab
DeleteCurrentSong = Delete
AddSongToPlaylist = Shift + A
ShowSongsInPlaylists = Shift + D
PlayOtherPlaylist = Shift + O
RedownloadCurrentSong = Shift + B
EditKeybindings = Shift + E
ChangeLanguage = Shift + L
PlayRandomSong = R
";

        public static int rewindSeconds = GetRewindSeconds();
        public static int forwardSeconds = GetForwardSeconds();
        public static float volume = GetVolume();
        public static float changeVolumeBy = GetChangeVolumeBy();
        public static float oldVolume = GetOldVolume();
        public static bool isLoop = GetIsLoop();
        public static bool isMuted = GetIsMuted();
        public static bool isShuffle = GetIsShuffle();
        public static bool isAutoSave = GetIsAutoSave();
        public static string localeLanguage = getLocaleLanguage();

        static public void CheckJammerFolderExists()
        {
            string jammerPath = Path.Combine(Utils.jammerPath);

            if (!Directory.Exists(jammerPath))
            {
                Directory.CreateDirectory(jammerPath);
                Directory.CreateDirectory(Path.Combine(jammerPath, "playlists"));
            }

            // check if settings.json exists but the file is empty, if so, delete it
            if (File.Exists(jammerPath + "/settings.json"))
            {
                string jsonString = File.ReadAllText(jammerPath + "/settings.json");
                if (jsonString == "")
                {
                    File.Delete(jammerPath + "/settings.json");
                }
            }
        }

        static public void SaveSettings()
        {
            string jammerPath = Path.Combine(Utils.jammerPath, "settings.json");
            // wirte hello world to a file
            Settings settings = new Settings();
            settings.IsLoop = isLoop;
            settings.Volume = volume;
            settings.isMuted = isMuted;
            settings.OldVolume = oldVolume;
            settings.forwardSeconds = forwardSeconds;
            settings.rewindSeconds = rewindSeconds;
            settings.changeVolumeBy = changeVolumeBy;
            settings.isShuffle = isShuffle;
            settings.isAutoSave = isAutoSave;
            settings.localeLanguage = localeLanguage;
            
            string jsonString = JsonSerializer.Serialize(settings);
            // delete file if exists
            if (File.Exists(jammerPath))
            {
                File.Delete(jammerPath);
            }
            File.WriteAllText(jammerPath, jsonString);
        }

        static public bool GetIsLoop()
        {
            string jammerPath = Path.Combine(Utils.jammerPath, "settings.json");
            if (File.Exists(jammerPath))
            {
                string jsonString = File.ReadAllText(jammerPath);
                Settings? settings = JsonSerializer.Deserialize<Settings>(jsonString);
                return settings?.IsLoop ?? false;
            }
            else
            {
                return false;
            }
        }

        static public float GetVolume()
        {
            string jammerPath = Path.Combine(Utils.jammerPath, "settings.json");
            if (File.Exists(jammerPath))
            {
                string jsonString = File.ReadAllText(jammerPath);
                Settings? settings = JsonSerializer.Deserialize<Settings>(jsonString);
                return settings?.Volume ?? 0.5f;
            }
            else
            {
                return 0.5f;
            }
        }

        static public bool GetIsMuted()
        {
            string jammerPath = Path.Combine(Utils.jammerPath, "settings.json");
            if (File.Exists(jammerPath))
            {
                string jsonString = File.ReadAllText(jammerPath);
                Settings? settings = JsonSerializer.Deserialize<Settings>(jsonString);
                return settings?.isMuted ?? false;
            }
            else
            {
                return false;
            }
        }

        static public float GetOldVolume()
        {
            string jammerPath = Path.Combine(Utils.jammerPath, "settings.json");
            if (File.Exists(jammerPath))
            {
                string jsonString = File.ReadAllText(jammerPath);
                Settings? settings = JsonSerializer.Deserialize<Settings>(jsonString);
                return settings?.OldVolume ?? 0.5f;
            }
            else
            {
                return 0.5f;
            }
        }

        static public int GetForwardSeconds()
        {
            string jammerPath = Path.Combine(Utils.jammerPath, "settings.json");
            if (File.Exists(jammerPath))
            {
                string jsonString = File.ReadAllText(jammerPath);
                Settings? settings = JsonSerializer.Deserialize<Settings>(jsonString);
                return settings?.forwardSeconds ?? 5;
            }
            else
            {
                return 5;
            }
        }

        static public int GetRewindSeconds()
        {
            string jammerPath = Path.Combine(Utils.jammerPath, "settings.json");
            if (File.Exists(jammerPath))
            {
                string jsonString = File.ReadAllText(jammerPath);
                Settings? settings = JsonSerializer.Deserialize<Settings>(jsonString);
                return settings?.rewindSeconds ?? 5;
            }
            else
            {
                return 5;
            }
        }

        static public float GetChangeVolumeBy()
        {
            string jammerPath = Path.Combine(Utils.jammerPath, "settings.json");
            if (File.Exists(jammerPath))
            {
                try {
                    string jsonString = File.ReadAllText(jammerPath);
                    Settings? settings = JsonSerializer.Deserialize<Settings>(jsonString);
                    return settings?.changeVolumeBy ?? 0.05f;
                }
                catch (Exception e)
                {
                    Console.WriteLine(e);
                    return 0.05f;
                }   
            }
            else
            {
                return 0.05f;
            }
        }

        static public bool GetIsShuffle()
        {
            string jammerPath = Path.Combine(Utils.jammerPath, "settings.json");
            if (File.Exists(jammerPath))
            {
                string jsonString = File.ReadAllText(jammerPath);
                Settings? settings = JsonSerializer.Deserialize<Settings>(jsonString);
                return settings?.isShuffle ?? false;
            }
            else
            {
                return false;
            }
        }
        static public string? getLocaleLanguage()
        {
            string jammerPath = Path.Combine(Utils.jammerPath, "settings.json");
            if (File.Exists(jammerPath))
            {
                string jsonString = File.ReadAllText(jammerPath);
                Settings? settings = JsonSerializer.Deserialize<Settings>(jsonString);
                return settings?.localeLanguage;
            }
            else
            {
                return "en";
            }
        }

        static public bool GetIsAutoSave()
        {
            string jammerPath = Path.Combine(Utils.jammerPath, "settings.json");
            if (File.Exists(jammerPath))
            {
                string jsonString = File.ReadAllText(jammerPath);
                Settings? settings = JsonSerializer.Deserialize<Settings>(jsonString);
                return settings?.isAutoSave ?? false;
            }
            else
            {
                return false;
            }
        }

        static public void OpenJammerFolder()
        {
            string jammerPath = Path.Combine(Utils.jammerPath, "settings.json");
            // start file managert in the given operating system
            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
            {
                System.Diagnostics.Process.Start("explorer.exe", jammerPath);
            }
            else if (RuntimeInformation.IsOSPlatform(OSPlatform.Linux))
            {
                System.Diagnostics.Process.Start("xdg-open", jammerPath);
            }
            else if (RuntimeInformation.IsOSPlatform(OSPlatform.OSX))
            {
                System.Diagnostics.Process.Start("open", jammerPath);
            }
        }

        public class Settings
        {
            public bool IsLoop { get; set; }
            public float Volume { get; set; }
            public float OldVolume { get; set; }
            public bool isMuted { get; set; }
            public int forwardSeconds { get; set; }
            public int rewindSeconds { get; set; }
            public float changeVolumeBy { get; set; }
            public bool isShuffle { get; set; }
            public bool isAutoSave { get; set; }
            public string? localeLanguage { get; set; }
            
        }
    }
}

File: src/preferences.py
import os
import json
import platform
import subprocess


def _get_setting(setting_name):
    jammer_path = os.path.join(
        os.path.expanduser("~"), "jammer", "settings.json")
    if os.path.exists(jammer_path):
        with open(jammer_path, "r") as file:
            settings = json.load(file)
            return settings[setting_name]
    else:
        return None
    
def get_is_loop():
    return _get_setting("IsLoop")

def get_volume():
    return _get_setting("Volume")

def get_is_muted():
    return _get_setting("isMuted")

def get_old_volume():
    return _get_setting("OldVolume")

def get_forward_seconds():
    return _get_setting("forwardSeconds")

def get_rewind_seconds():
    return _get_setting("rewindSeconds")

def get_change_volume_by():
    return _get_setting("changeVolumeBy")

def get_is_shuffle():
    return _get_setting("isShuffle")

# Define default values
is_loop = get_is_loop() or False
volume = get_volume() or 50
is_muted = get_is_muted() or False
old_volume = get_old_volume() or 0
forward_seconds = get_forward_seconds() or 10
rewind_seconds = get_rewind_seconds() or 10
change_volume_by = get_change_volume_by() or 5
is_shuffle = get_is_shuffle() or False

def check_jammer_folder_exists():
    jammer_path = os.path.join(os.path.expanduser("~"), "jammer")
    if not os.path.exists(jammer_path):
        os.makedirs(jammer_path)
        os.makedirs(os.path.join(jammer_path, "playlists"))

    settings_path = os.path.join(jammer_path, "settings.json")
    if os.path.exists(settings_path):
        with open(settings_path, "r") as file:
            if file.read() == "":
                set_settings()
                save_settings()
    else:
        set_settings()
        save_settings()

def save_settings():
    jammer_path = os.path.join(
        os.path.expanduser("~"), "jammer", "settings.json")

    settings = {
        "IsLoop": is_loop,
        "Volume": volume,
        "isMuted": is_muted,
        "OldVolume": old_volume,
        "forwardSeconds": forward_seconds,
        "rewindSeconds": rewind_seconds,
        "changeVolumeBy": change_volume_by,
        "isShuffle": is_shuffle
    }

    with open(jammer_path, "w") as file:
        json.dump(settings, file)

def set_settings():
    global is_loop, volume, is_muted, old_volume, forward_seconds, rewind_seconds, change_volume_by, is_shuffle
    is_loop = False 
    volume = 50
    is_muted = False
    old_volume = 0
    forward_seconds = 10
    rewind_seconds = 10
    change_volume_by = 5
    is_shuffle = False
    
def open_jammer_folder():
    jammer_path = os.path.join(os.path.expanduser("~"), "jammer")

    if platform.system() == "Windows":
        subprocess.Popen(["explorer.exe", jammer_path], shell=True)
    elif platform.system() == "Linux":
        subprocess.Popen(["xdg-open", jammer_path])
    elif platform.system() == "Darwin":
        subprocess.Popen(["open", jammer_path])
